File: src/steam/vdf.cpp
#include "vdf.h"
#include <QFile>
#include <QTextStream>
#include <QSet>
#include <QVector>
#include <cctype>

namespace ProtonSage {

// ── Lexer / Scanner ─────────────────────────────────────────────────

namespace {

enum class TokenKind { Eof, String, LBrace, RBrace };

struct Token {
    TokenKind kind = TokenKind::Eof;
    QString   text;
};

class Scanner {
public:
    explicit Scanner(const QByteArray& data)
        : m_data(QString::fromUtf8(data))
        , m_pos(0)
        , m_line(1)
        , m_col(1)
    {}

    Token nextToken();

private:
    void skipWhitespaceAndComments();
    Token scanQuoted();
    Token scanBare();
    QChar peek() const;
    QChar advance();

    QString m_data;
    int     m_pos = 0;
    int     m_line = 1;
    int     m_col  = 1;
};

QChar Scanner::peek() const
{
    if (m_pos >= m_data.size())
        return QChar();
    return m_data.at(m_pos);
}

QChar Scanner::advance()
{
    QChar ch = m_data.at(m_pos);
    ++m_pos;
    if (ch == QLatin1Char('\n')) {
        ++m_line;
        m_col = 1;
    } else {
        ++m_col;
    }
    return ch;
}

void Scanner::skipWhitespaceAndComments()
{
    while (m_pos < m_data.size()) {
        QChar ch = peek();
        if (ch.isSpace()) {
            advance();
            continue;
        }
        // C++ style // comments (Steam VDF uses these sometimes)
        if (ch == QLatin1Char('/') && m_pos + 1 < m_data.size()
            && m_data.at(m_pos + 1) == QLatin1Char('/')) {
            while (m_pos < m_data.size() && peek() != QLatin1Char('\n'))
                advance();
            continue;
        }
        break;
    }
}

Token Scanner::scanQuoted()
{
    advance(); // consume opening '"'
    QString buf;
    while (m_pos < m_data.size()) {
        QChar ch = advance();
        if (ch == QLatin1Char('"'))
            return Token{TokenKind::String, buf};
        if (ch == QLatin1Char('\\') && m_pos < m_data.size()) {
            QChar next = advance();
            switch (next.unicode()) {
            case 'n': buf += QLatin1Char('\n'); break;
            case 't': buf += QLatin1Char('\t'); break;
            case 'r': buf += QLatin1Char('\r'); break;
            default:  buf += next; break;
            }
            continue;
        }
        buf += ch;
    }
    // Unterminated string — return what we have
    return Token{TokenKind::String, buf};
}

Token Scanner::scanBare()
{
    QString buf;
    while (m_pos < m_data.size()) {
        QChar ch = peek();
        if (ch.isSpace() || ch == QLatin1Char('{') || ch == QLatin1Char('}'))
            break;
        buf += advance();
    }
    return Token{TokenKind::String, buf};
}

Token Scanner::nextToken()
{
    skipWhitespaceAndComments();
    if (m_pos >= m_data.size())
        return Token{TokenKind::Eof, QString()};

    QChar ch = peek();
    switch (ch.unicode()) {
    case '{':
        advance();
        return Token{TokenKind::LBrace, QStringLiteral("{")};
    case '}':
        advance();
        return Token{TokenKind::RBrace, QStringLiteral("}")};
    case '"':
        return scanQuoted();
    default:
        return scanBare();
    }
}

// ── Parser ──────────────────────────────────────────────────────────

class Parser {
public:
    explicit Parser(const QByteArray& data)
        : m_scanner(data)
    {}

    VDFObject parse();

private:
    VDFObject parseObject(bool stopOnRBrace);

    Scanner m_scanner;
};

VDFObject Parser::parse()
{
    return parseObject(false);
}

VDFObject Parser::parseObject(bool stopOnRBrace)
{
    VDFObject obj;
    while (true) {
        Token keyTok = m_scanner.nextToken();

        switch (keyTok.kind) {
        case TokenKind::Eof:
            if (stopOnRBrace)
                return VDFObject(); // error: unexpected EOF inside object
            return obj;
        case TokenKind::RBrace:
            if (stopOnRBrace)
                return obj;
            return VDFObject(); // error: unexpected }
        case TokenKind::LBrace:
            return VDFObject(); // error: unexpected {
        case TokenKind::String:
            break; // continue below
        }

        Token valueTok = m_scanner.nextToken();
        switch (valueTok.kind) {
        case TokenKind::String:
            obj.insert(keyTok.text, QVariant(valueTok.text));
            break;
        case TokenKind::LBrace: {
            VDFObject child = parseObject(true);
            if (child.isEmpty() && !stopOnRBrace) {
                // Could be an error, but still store empty object
            }
            obj.insert(keyTok.text, QVariant::fromValue(child));
            break;
        }
        case TokenKind::Eof:
            return VDFObject(); // error: missing value
        case TokenKind::RBrace:
            return VDFObject(); // error: missing value before }
        }
    }
}

} // anonymous namespace

// ── Public API ───────────────────────────────────────────────────────

VDFObject parseVDF(const QByteArray& data)
{
    Parser parser(data);
    return parser.parse();
}

VDFObject parseVDFFile(const QString& path)
{
    QFile file(path);
    if (!file.open(QIODevice::ReadOnly))
        return VDFObject();

    QByteArray data = file.readAll();
    file.close();
    return parseVDF(data);
}

// ── Helper implementations ──────────────────────────────────────────

bool asObject(const QVariant& value, VDFObject& out)
{
    if (!value.isValid())
        return false;

    // QVariantMap is the type used for VDFObject
    if (value.canConvert<QVariantMap>()) {
        out = value.toMap();
        return true;
    }
    return false;
}

QString stringValue(const VDFObject& obj, const QString& key)
{
    auto it = obj.constFind(key);
    if (it == obj.constEnd())
        return QString();

    if (it->isValid() && it->canConvert<QString>())
        return it->toString();

    return QString();
}

QString stringValueFold(const VDFObject& obj, const QStringList& keys)
{
    // Exact match first
    for (const QString& key : keys) {
        auto it = obj.constFind(key);
        if (it != obj.constEnd() && it->canConvert<QString>()) {
            QString value = it->toString().trimmed();
            if (!value.isEmpty())
                return value;
        }
    }

    // Case-insensitive fallback
    for (const QString& key : keys) {
        for (auto it = obj.constBegin(); it != obj.constEnd(); ++it) {
            if (it.key().compare(key, Qt::CaseInsensitive) == 0
                && it->canConvert<QString>()) {
                QString value = it->toString().trimmed();
                if (!value.isEmpty())
                    return value;
            }
        }
    }

    return QString();
}

bool objectValueFold(const VDFObject& obj, const QString& key, VDFObject& out)
{
    // Exact match
    auto it = obj.constFind(key);
    if (it != obj.constEnd() && asObject(it.value(), out))
        return true;

    // Case-insensitive fallback
    for (auto it = obj.constBegin(); it != obj.constEnd(); ++it) {
        if (it.key().compare(key, Qt::CaseInsensitive) == 0
            && asObject(it.value(), out))
            return true;
    }

    return false;
}

} // namespace ProtonSage

File: internal/steam/vdf.go
package steam

import (
	"fmt"
	"os"
	"strings"
	"unicode"
)

// VDFObject is a minimal Valve KeyValues object representation.
type VDFObject map[string]any

type tokenKind int

const (
	tokenEOF tokenKind = iota
	tokenString
	tokenLBrace
	tokenRBrace
)

type token struct {
	kind tokenKind
	text string
}

type scanner struct {
	input []rune
	pos   int
	line  int
	col   int
}

func newScanner(data []byte) *scanner {
	return &scanner{input: []rune(string(data)), line: 1, col: 1}
}

func (s *scanner) nextToken() (token, error) {
	s.skipWhitespaceAndComments()
	if s.pos >= len(s.input) {
		return token{kind: tokenEOF}, nil
	}

	ch := s.peek()
	switch ch {
	case '{':
		s.advance()
		return token{kind: tokenLBrace, text: "{"}, nil
	case '}':
		s.advance()
		return token{kind: tokenRBrace, text: "}"}, nil
	case '"':
		return s.scanQuoted()
	default:
		return s.scanBare(), nil
	}
}

func (s *scanner) skipWhitespaceAndComments() {
	for s.pos < len(s.input) {
		ch := s.peek()
		if unicode.IsSpace(ch) {
			s.advance()
			continue
		}
		if ch == '/' && s.pos+1 < len(s.input) && s.input[s.pos+1] == '/' {
			for s.pos < len(s.input) && s.peek() != '\n' {
				s.advance()
			}
			continue
		}
		break
	}
}

func (s *scanner) scanQuoted() (token, error) {
	startLine, startCol := s.line, s.col
	s.advance() // opening quote
	var b strings.Builder
	for s.pos < len(s.input) {
		ch := s.advance()
		if ch == '"' {
			return token{kind: tokenString, text: b.String()}, nil
		}
		if ch == '\\' && s.pos < len(s.input) {
			next := s.advance()
			switch next {
			case 'n':
				b.WriteRune('\n')
			case 't':
				b.WriteRune('\t')
			case 'r':
				b.WriteRune('\r')
			default:
				b.WriteRune(next)
			}
			continue
		}
		b.WriteRune(ch)
	}
	return token{}, fmt.Errorf("unterminated quoted string at %d:%d", startLine, startCol)
}

func (s *scanner) scanBare() token {
	var b strings.Builder
	for s.pos < len(s.input) {
		ch := s.peek()
		if unicode.IsSpace(ch) || ch == '{' || ch == '}' {
			break
		}
		b.WriteRune(s.advance())
	}
	return token{kind: tokenString, text: b.String()}
}

func (s *scanner) peek() rune {
	return s.input[s.pos]
}

func (s *scanner) advance() rune {
	ch := s.input[s.pos]
	s.pos++
	if ch == '\n' {
		s.line++
		s.col = 1
	} else {
		s.col++
	}
	return ch
}

type vdfParser struct {
	s *scanner
}

// ParseVDF parses the subset of Valve KeyValues used by Steam library and app manifest files.
func ParseVDF(data []byte) (VDFObject, error) {
	p := &vdfParser{s: newScanner(data)}
	obj, err := p.parseObject(false)
	if err != nil {
		return nil, err
	}
	return obj, nil
}

// ParseVDFFile parses a VDF/ACF file from disk. It only reads the file.
func ParseVDFFile(path string) (VDFObject, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return ParseVDF(data)
}

func (p *vdfParser) parseObject(stopOnRBrace bool) (VDFObject, error) {
	obj := VDFObject{}
	for {
		keyTok, err := p.s.nextToken()
		if err != nil {
			return nil, err
		}

		switch keyTok.kind {
		case tokenEOF:
			if stopOnRBrace {
				return nil, fmt.Errorf("unexpected EOF inside object")
			}
			return obj, nil
		case tokenRBrace:
			if stopOnRBrace {
				return obj, nil
			}
			return nil, fmt.Errorf("unexpected } at top level")
		case tokenLBrace:
			return nil, fmt.Errorf("unexpected { before key")
		case tokenString:
			// Continue below.
		}

		valueTok, err := p.s.nextToken()
		if err != nil {
			return nil, err
		}
		switch valueTok.kind {
		case tokenString:
			obj[keyTok.text] = valueTok.text
		case tokenLBrace:
			child, err := p.parseObject(true)
			if err != nil {
				return nil, err
			}
			obj[keyTok.text] = child
		case tokenEOF:
			return nil, fmt.Errorf("missing value for key %q", keyTok.text)
		case tokenRBrace:
			return nil, fmt.Errorf("missing value for key %q before }", keyTok.text)
		}
	}
}

// AsObject casts a VDF value to an object.
func AsObject(value any) (VDFObject, bool) {
	obj, ok := value.(VDFObject)
	return obj, ok
}

// StringValue returns a string field from a VDF object.
func StringValue(obj VDFObject, key string) string {
	value, ok := obj[key]
	if !ok {
		return ""
	}
	str, _ := value.(string)
	return str
}
